### ISSUE TITLE
optimising prosolo performance

### ISSUE THREAD

**Issue by dlaehnemann** [ProSolo/prosolo#2]
As `ProSolo` is looking at Single Nucleotide Variants (SNVs), this means inspecting a lot more positions than `prosic` previously did. While scaling linearly with the amount of sequencing data, we needed to improve the previous runtime.

The first step was to flamegraph the running of the tool before optimisations (for methods see below). As it is not straightforward to include SVGs in GitHub issues, I'll send the respective files via eMail.

So I created an initial flamegraph with `perf + flamegraph.pl` and inspected it with @johanneskoester. The main things that stuck out were bigger chunks of time being spent in `bam_aux_get()`, an `htslib` functionality for getting auxiliairy read data, related calls to `C` string creation functions and a lot of time spent in the likelihood calculations in log space. Thus, the main ideas we came up with were:

1. **Ensure that each read and each piece of auxiliary data is read only once, and as the `time` analysis below shows, the changes implemented by @johanneskoester heavily improve runtime by 1/3!**
2. Ignore reads in the current read buffer that start right of a candidate SNV, in contrast to candidate indels where these might be informative. This change implemented by @johanneskoester did not cause any measurable speedup.
3. Avoiding the recalculation of likelihoods for single cell sample and bulk sample allele frequencies for different `Event`s where they might be needed. This change that I implemented, did not yield any measurable speedup. My assumption is that the reduced number of likelihood calculations is offset by a slight increase in overhead caused by the caching.

To validate these findings based on timing and to see whether any further optimisations might make sense, I then also created more detailed flamegraphs with `rust` functionality. This demonstrated nicely that the first change strongly reduced the CIGAR string related time chunks, whereas the other changes did not noticeably affect relative time usage of different code parts. In addition, the resulting flamegraphs after all optimisations clearly show, that most of the time is now spent in calls to low-level rust library functions (`rust-bio` log likelihood calculations) and eventually system library log calculation functions. My conclusion for now would thus be:

We are down to pretty much the most efficient code we can get with `libprosic` and `prosolo`, and all we can do now to reduce runtime would mean reducing the amount of data we look at. I.e. we would have to introduce heuristics to downsample single cell or bulk read coverage at certain sites. To do this, we would have to track how any of those changes affect especially the recall, as throwing away data bears the danger of throwing away alternative allele evidence.

# baseline timing

Before we started on optimisations, `time` for the flamegraphing run gave the following runtime:
```
real    35m32.092s
user    35m30.020s
sys 0m2.208s
```
A second run confirmed the runtime:
```
real    35m4.222s
user    35m2.892s
sys 0m1.172s
```

# unpack CIGAR Strings only once

https://github.com/PROSIC/libprosic/pull/34

Runtime after this optimisation decreased by about a third:
```
real    23m16.300s
user    23m14.684s
sys 0m1.028s
```

# ignore reads right of SNVs

https://github.com/PROSIC/libprosic/commit/72bd0d7dc3f9da9976284bdb8f8e72caba22233e

Runtime after this optimisation did improve significantly:
```
real    22m58.577s
user    22m57.604s
sys 0m0.868s
```

# caching of likelihoods

https://github.com/PROSIC/libprosic/pull/36

Runtime after this optimisation did not improve significantly:
```
real    23m9.813s
user    23m8.716s
sys 0m0.996s
```

# flamegraph procedure

### One-time setup

1. Install `perf`. Under Ubuntu, apparently the wrapper packages `linux-tools-generic` and `linux-cloud-tools-generic` do not necessarily do the trick -- you will sometimes have to explicitly request the ones for your kernel version by using:
```
sudo apt install linux-tools-generic linux-cloud-tools-generic linux-tools-`uname -r`-generic linux-cloud-tools-`uname -r`-generic
```
Or you might even have to explicitly ask for an older version to meet dependencies. Trying to run perf I got an error message I cannot reconstruct any more, but from it I deduced trying to install the following version and it worked:
```
sudo apt install linux-tools-generic linux-cloud-tools-generic linux-tools-4.4.0-127-generic linux-cloud-tools-4.4.0-127-generic
```
2. Install `flamegraph` from repo:
  1. Clone the repo locally: `git clone https://github.com/brendangregg/FlameGraph`
  2. Add the main directory withh all the `*.pl` Perl files to the path.

### flamegraph a binary

To turn on debugging information in the binary, to get actual function names in the flamegraph output, temporarily add to `Cargo.toml`:

```
[profile.release]
debug = true
```

Then compile with the `--release` flag, to get cargo to optimize the resulting binary. Otherwise, any slowness may be due to a lack of compiler optimisations:

```
cargo build --release
```

Run the cpu sampling with:

```
perf record -g -F 10000 ../target/release/name-of-binary <command-line-arguments>
```

In our case, a good example run from a subdirectory of my local `prosolo/benches` directory was:
```
time perf record -g -F 10000 ../../target/release/bins_7dfda4ef909015a3fb31557bfc0ab7a2b29b7386/prosolo single-cell-bulk --omit-indels --candidates ../PAG5.PNG.prosolo_candidates.chr4_3000000-4000000.bcf ../PAG5.bps.sorted.chr4_3000000-4000000.bam ../PNG.bps.sorted.chr4_3000000-4000000.bam ~/analysis/hg19/ucsc.hg19.fasta  --sc-isize-mean 202.8 --sc-isize-sd 67.2 --output PAG5.chr4_3000000-4000000.prosolo.bcf
```

The resulting `report.perf` can be rather large, depending on the length of your example run and the sampling frequency selected -- e.g. 1G for a half hour run of `prosolo` with the above frequency. Based on the report, generate the flamegraph with:

```
perf script | stackcollapse-perf.pl | flamegraph.pl > flame.svg
```

Inspect the `flame.svg` file by opening it in a browser and hovering over individual bars to get the respective function names displayed.

Or inspect the report interactively by issuing:
```
perf report
```

## Rust internal flamegraphing: flame and flamer crates

The crates used are (basic usage is documented on GitHub):
* `flame` provides flamegraphing functions for use within Rust code: https://github.com/TyOverby/flame
* `flamer` makes it easy to deploy via feature annotations: https://github.com/llogiq/flamer

Using [rustup](https://rustup.rs/), install the nightly Rust build:

```
rustup toolchain install nightly
```

Follow the [documentation of `flamer`](https://github.com/llogiq/flamer) to get going. In addition, add the following to your `main.rs` (or `lib.rs` in a library crate) -- it ensures that calls to the same function at the same level of the hierarchy are collapsed into one flamegraph bar (this ensures that your browser can cope with the graph and makes it a more useful representation than with thousands to millions of little blocks of the same name):
```
/// Implementation as given by @ebarnard at: https://github.com/TyOverby/flame/issues/33#issuecomment-352312506
#[cfg(feature="flame_it")]
fn write_flamegraph() {
    let mut spans = flame::threads().into_iter().next().unwrap().spans;
    merge_spans(&mut spans);
    flame::dump_html_custom(&mut File::create("flame-graph.html").unwrap(), &spans)
        .unwrap();
}

fn merge_spans(spans: &mut Vec<flame::Span>) {
    if spans.is_empty() {
        return;
    }

    // Sort so spans to be merged are adjacent and spans with the most children are
    // merged into to minimise allocations.
    spans.sort_unstable_by(|s1, s2| {
        let a = (&s1.name, s1.depth, usize::max_value() - s1.children.len());
        let b = (&s2.name, s2.depth, usize::max_value() - s2.children.len());
        a.cmp(&b)
    });

    // Copy children and sum delta from spans to be merged
    let mut merge_targets = vec![0];
    {
        let mut spans_iter = spans.iter_mut().enumerate();
        let (_, mut current) = spans_iter.next().unwrap();
        for (i, span) in spans_iter {
            if current.name == span.name && current.depth == span.depth {
                current.delta += span.delta;
                let mut children = mem::replace(&mut span.children, Vec::new());
                current.children.extend(children.into_iter());
            } else {
                current = span;
                merge_targets.push(i);
            }
        }
    }

    // Move merged spans to the front of the spans vector
    for (target_i, &current_i) in merge_targets.iter().enumerate() {
        spans.swap(target_i, current_i);
    }

    // Remove duplicate spans
    spans.truncate(merge_targets.len());

    // Merge children of the newly collapsed spans
    for span in spans {
        merge_spans(&mut span.children);
    }
}


#[cfg(not(feature="flame_it"))]
fn write_flamegraph() {
}
```

And add `write_flamegraph();` to the end of your `main()` function. Then, build with the feature `flame_it` activated and using the nightly build:

```
rustup run nightly cargo build --release --features flame_it
```

Using the generated binary, run the program on some example data, in my case:
```
../../target/release/bin_flame_it/prosolo single-cell-bulk --omit-indels --candidates ../PAG5.PNG.prosolo_candidates.chr7_100000-200000.bcf ../PAG5.bps.sorted.chr7_100000-200000.bam ../PNG.bps.sorted.chr7_100000-200000.bam ~/analysis/hg19/ucsc.hg19.fasta  --sc-isize-mean 196.1 --sc-isize-sd 71.1 --output PAG5.chr7_100000-200000.prosolo.bcf
```

And inspect...

**Comment by alicemchardy** [comment]
Thanks for the nice summary. Can you specify dataset size and system used for calculations?

**Comment by dlaehnemann** [comment]
Yes, good point. All of this ran on my notebook on an individual core (`Intel® Core™ i7-6600U CPU @ 2.60GHz`) and was looking at the exome of one human cell and the respective high coverage bulk, inspecting only a 1 Mb piece of one chromosome, i.e. `chr4:3000000-4000000`.

Only for one of the later flamegraphs that you received via eMail (`06_flame_it_detail_chr7_flame-graph.html`) I used an even shorter 1 kb fragment of `chr7:100000-200000`, because the more detailed flamegraphing quickly blew up memory usage on the larger segment.

**Comment by johanneskoester** [comment]
Hi all,
I just looked at your code for the likelihood caching again and found [this](https://github.com/PROSIC/libprosic/blob/master/src/model/mod.rs#L536) code line.

The line means that you insert the likelihood into the cache if you cannot find it for the given allele frequency. So far so good. However, you use the method `or_insert` for that, which takes the value to insert as an argument. When this is executed, it means that, in any case, the value has to be calculated, in order to be passed to the method. Also if it is actually not needed because the allele frequency is found.
What you have to do instead to benefit from the caching is to only calculate that value lazily if you cannot find the allele frequency. The function for that is `or_insert_with`, which takes a function, that is only executed in the "or" case. See [here](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.or_insert_with)

Sorry that I only see that now and not earlier while reviewing your PR. I hope your analysis is automated enough to not have to redo a lot of work after changing this. I would think that afterwards the caching should give you a significant speedup.

Best,
Johannes

---
https://koesterlab.github.io

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by dlaehnemann** [comment]
Aaah, thanks for catching this now! ;)

I was wondering why it didn't give any improvement, but this sounds promising. Will check this out!

**Comment by dlaehnemann** [comment]
# caching of likelihoods patched

With the mistake that Johannes caught corrected, we get a runtime of:
```
real	4m11.441s
user	4m10.248s
sys	0m0.944s
```

So we save about 80% on the previous best runtime!

Will post the respective commit once it's pushed.

**Comment by johanneskoester** [comment]
Great! Can you send us an updated flamegraph (with the rust method)? We can use that to see if any bottleneck remains.

**Comment by dlaehnemann** [comment]
Jep, it's in the making. @aschoen has also already asked for one. I have the `perf` generated one already, but am compiling the `flame_it` binary as I write. Will send both flamegraphs around via eMail later.

But I think the current version is also good enough to get on with comparative benchmarking, for now. We should maybe bundle everything up in proper library releases (`rust-htslib`, `rust-bio`, `libprosic`) for a proper new `ProSolo` release?

**Comment by alicemchardy** [comment]
Sounds good!

**Comment by dlaehnemann** [comment]
Further optimisations in libprosic give the following ProSolo speedups (always compared to the latest version right before the PR):

* caching of the `prob_rho()` function through [libprosic PR 46](https://github.com/PROSIC/libprosic/pull/46) gives a further speedup of about 20%
* storing mismapping along with the mapping probability per `Observation` (=read) through [libprosic PR 48](https://github.com/PROSIC/libprosic/pull/48) gives a further speedup of about 15%

**Comment by dlaehnemann** [comment]
@johanneskoester implemented a bunch of further optimisations mainly aimed at `prosic` into `libprosic`. `prosolo` using `libprosic 0.7.2` now runs the same data with:

```
user	2m7.432s
user	2m6.180s
user	2m5.476s
user	2m6.124s

((2+2+2+2)*60 + 7.432 + 6.180 + 5.476 + 6.124)/4 = 126,303
```

Compared to the last timing in <https://github.com/PROSIC/libprosic/pull/48>, this makes a further reduction in runtime of `prosolo` of:
```
(163,702 - 126,303) / 163,702 * 100 % = 23 %
```

For `prosic`, these changes improve runtime more drastically, but `prosolo` profits noticeably, as well!

**Comment by dlaehnemann** [comment]
I'm closing this for now, as we have spent quite a lot of energy optimising parts of libprosic. For `prosolo`, we have probably have exhausted most of the major possible optimisation, so unless we come up with new ideas or the need to further speed up `prosolo`, I think the goal of this issue has been accomplished.
